sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ReadFromKafkaViaSDF.java

```diff
@@ -0,0 +1,697 @@
+ * .apply(ReadFromKafkaViaSDF.create()
+ *          .withBootstrapServers("broker_1:9092,broker_2:9092")
+ *          .withKeyDeserializer(LongDeserializer.class).
+ *          .withValueDeserializer(StringDeserializer.class)
+ *          .withProcessingTime()
+ *          .commitOffsets());
+ *
+ * }</pre>
+ *
+ * <h3>Read from {@link KafkaSourceDescription}</h3>
+ *
+ * {@link ReadFromKafkaDoFn} implements the logic of reading from Kafka. The element is a {@link
+ * KafkaSourceDescription}, and the restriction is an {@link OffsetRange} which represents record
+ * offset. A {@link GrowableOffsetRangeTracker} is used to track an {@link OffsetRange} ended with
+ * {@code Long.MAX_VALUE}. For a finite range, a {@link OffsetRangeTracker} is created.
+ *
+ * <h4>Initialize Restriction</h4>
+ *
+ * {@link ReadFromKafkaDoFn#initialRestriction(KafkaSourceDescription)} creates an initial range for
+ * a input element {@link KafkaSourceDescription}. The end of range will be initialized as {@code
+ * Long.MAX_VALUE}. For the start of the range:
+ *
+ * <ul>
+ *   <li>If {@link KafkaSourceDescription#getStartOffset()} is set, use this offset as start.
+ *   <li>If {@link KafkaSourceDescription#getStartReadTime()} is set, seek the start offset based on
+ *       this time.
+ *   <li>Otherwise, the last committed offset + 1 will be returned by {@link
+ *       Consumer#position(TopicPartition)} as the start.
+ * </ul>
+ *
+ * <h4>Initial Split</h4>
+ *
+ * <p>There is no initial split for now.
+ *
+ * <h4>Checkpoint and Resume Processing</h4>
+ *
+ * <p>There are 2 types of checkpoint here: self-checkpoint which invokes by the DoFn and
+ * system-checkpoint which is issued by the runner via {@link
+ * org.apache.beam.model.fnexecution.v1.BeamFnApi.ProcessBundleSplitRequest}. Every time the
+ * consumer gets empty response from {@link Consumer#poll(long)}, {@link ReadFromKafkaDoFn} will
+ * checkpoint at current {@link KafkaSourceDescription} and move to process the next element. These
+ * deferred elements will be resumed by the runner as soon as possible.
+ *
+ * <h4>Progress and Size</h4>
+ *
+ * <p>The progress is provided by {@link GrowableOffsetRangeTracker} or {@link OffsetRangeTracker}
+ * per {@link KafkaSourceDescription}. For an infinite {@link OffsetRange}, a Kafka {@link Consumer}
+ * is used in the {@link GrowableOffsetRangeTracker} as the {@link
+ * GrowableOffsetRangeTracker.RangeEndEstimator} to poll the latest offset. Please refer to {@link
+ * ReadFromKafkaDoFn.KafkaLatestOffsetEstimator} for details.
+ *
+ * <p>The size is computed by {@link ReadFromKafkaDoFn#getSize(KafkaSourceDescription,
+ * OffsetRange).} A {@link KafkaIOUtils.MovingAvg} is used to track the average size of kafka
+ * records.
+ *
+ * <h4>Track Watermark</h4>
+ *
+ * The estimated watermark is computed by {@link MonotonicallyIncreasing} based on output timestamps
+ * per {@link KafkaSourceDescription}.
+ */
+@AutoValue
+public abstract class ReadFromKafkaViaSDF<K, V>
```

Comment:
This seems strangely close to something we lived in the SDF version of HBaseIO. In the first version we did an artificial object called `HBaseQuery` that contained the minimum information we needed to be able to query the Data store in a SDF way, but then other requirements came in and we started to add extra parameters to end up with something that was almost close to the exact 'complete' specification of the Read class so we decided to read from a PCollection<Read> otherwise we will be duplicating code, so we ended up with https://github.com/apache/beam/blob/f6ef9032f521180f1cc26959d9d6ab86dd37a13c/sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java#L353

Here you can have `PCollection<Read>` as an input and get rid of `KafkaSourceDescription` and this will have a more consistent user experience for final users. Notice that this `ReadAll` like pattern is also now used now in [SolrIO](https://github.com/apache/beam/blob/f6ef9032f521180f1cc26959d9d6ab86dd37a13c/sdks/java/io/solr/src/main/java/org/apache/beam/sdk/io/solr/SolrIO.java#L501) and there is an ongoing PR to introduce it for [CassandraIO](https://github.com/apache/beam/pull/10546) so maybe it is a good idea we follow it for consistency.

Notice that in the SolrIO case the change looks even closer to this one because we ended up introducing `ReplicaInfo` (the spiritual equivalent of `TopicPartition`) into normal Read and we guarantee in expansion that this field gets filled if the users don't do it, but if they do well we asume they know what they are doing and we go with it.

Another advantage of having the full specification is that you will be able to read not only from multiple topics but also from different clusters because of the power of having the full `Read` specification,

Reply:
I'm agree with @iemejia in the way that `HbaseIO` is a good example when we can end up with reimplementing new `Read` in the dedicated initially more light class, like `KafkaSourceDescription`, in the end. 
1. Yes, it's quite heavy but we don't expect too many elements like this in the PCollection, right? So it should not be a problem.
2. Of course, reading from several clusters is not a main feature but, afair, we had several requests on that from users.
3. Agree, this is a good point. Do we have any principal objections to use `Read` for that?

Reply:
Sorry for the late. 

> 3. Agree, this is a good point. Do we have any principal objections to use Read for that?

Yes, there are some concerns around that. As I mentioned in point1, `KafkaIO.Read` contains lots of configurations which are supported to determined during runtime. For these part, they shouldn't be the element logically. Also `KafkaIO.Read` is also used by the user who doesn't migrate to fnapi yet, which may last for a long time. It will not be easy for us to maintain or develop new things based on `KafkaIO.Read` when considering compatibility.

I think we should figure out the major needs from KafkaIO beam user, especially what they want to be dynamic during pipeline runtime. Alexey, do you know what can be a good place to collect such idea?

Reply:
I'm not worried about the size of the `source descriptor` for Kafka but more about the pattern in which people would build and use these objects and about the backwards/forwards compatibility in a pipeline update scenario. I think it would be best if we used a dedicated object not extending the `PTransform` because:
1) it simplifies the migration scenario to a 100% SDF world. `Read` becomes a wrapper around `Create(source descriptor) -> ReadAll` when using SDF otherwise we use the existing `UnboundedSource` implementation and it pulls parameters off of this `source descriptor`
2) we don't couple source descriptor with classes/evolution of pipeline construction APIs
3) seems awkward from a user perspective to have a PCollection<PTransform> -> PTransform -> PCollection<KafkaRecord>

Reply:
> I'm agree with @iemejia in the way that `HbaseIO` is a good example when we can end up with reimplementing new `Read` in the dedicated initially more light class, like `KafkaSourceDescription`, in the end.
> 
> 1. Yes, it's quite heavy but we don't expect too many elements like this in the PCollection, right? So it should not be a problem.
> 2. Of course, reading from several clusters is not a main feature but, afair, we had several requests on that from users.
> 3. Agree, this is a good point. Do we have any principal objections to use `Read` for that?

Hi Alexey, I want to elaborate  more details on having bootstrapServer dynamically. I'm wondering what should we do if the user sets bootstrapServer in the consumer config but also emits different bootstrapServer dynamically. In this case, should we override the consumer config with the dynamic one or we just update the consumer config with the new one?

Reply:
@boyuanzz It's quite tricky question. Well, initially we use the static list of bootstrap servers for message and offset consumers. So we expect them equal and I think it should be the same with SDF.
Message consumer is also used to fetch the topic(s) partitions for initial split. With SDF we don't need that. 
What is not clear for me, is by "user emits different bootstrapServer dynamically" you mean that it will be set in a runtime? Is it going to be changed during a runtime or set only once?

Reply:
The question here is, do we want to only have bootstrap servers in element during runtime, or we want to allow the user to set the bootstrap server when constructing the pipeline, and mean while also allow the user to provide new bootstrap server during runtime. For example, if the customer gives the static list of [bootstrap1] during construction time, and also give new bootstrap2 in the runtime, what should be the list we use to connect:
1. just use [bootstrap1]
2. or just use [bootstrap2]
3. or [bootstrap1, bootstrap2]

Reply:
I'd suggest to stick with 1 (at least for now). 2 and 3 seems to me could be an error prone since I expect it will require much more work in case if the new servers will be related to different cluster, for example. Wdyt?

Reply:
Since we want to empower the user to have bootstrapServer on the fly, I think we we have 2 options:
1. We remove the `bootstrapServers` from construction time configuration and `KafkaSourceDescription` will be the source of truth.
2. We still allow setting default `bootstrapServers` at construction time, and treat `KafkaSourceDescription.bootstrapServers` as override. 

I'm leaning to #2 as long as it's easy for the user to understand the meaning of override.

Reply:
Agree with option #2 as well.

Reply:
Thanks! I'll ping this thread once the PR is ready for another round of review.